### PR TITLE
aws_kms: Update policy on existing keys (when passed)

### LINKED ISSUE #59987: aws_kms doesn't update policy
##### SUMMARY

If you attempt to modify the policy on an existing KMS CMK using aws_kms's "policy" option this doesn't appear to get updated.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/aws_kms.py

##### ANSIBLE VERSION

```
ansible 2.8.1
  config file = /home/mchappel/.ansible.cfg
  configured module search path = [u'/home/mchappel/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jun 11 2019, 12:19:05) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
```

##### CONFIGURATION

```
DEFAULT_CALLBACK_WHITELIST(/home/mchappel/.ansible.cfg) = [u'timer', u'profile_tasks']
HOST_KEY_CHECKING(/home/mchappel/.ansible.cfg) = False
RETRY_FILES_ENABLED(/home/mchappel/.ansible.cfg) = False
```

##### OS / ENVIRONMENT
RHEL7 + AWS

##### STEPS TO REPRODUCE

- Create KMS CMK
- Attempt to update policy using the 'policy' option on aws_kms

```
    - name: 'Update policy on KMS Keys'
      aws_kms:
        key_id: '{{ kms_key.id }}'
        alias: '{{ kms_key.alias | default(omit) }}'
        policy: ' {{ kms_current_policy | to_json }}'
```

##### EXPECTED RESULTS
KMS Policy is updated

##### ACTUAL RESULTS
KMS Policy stays the same

### PR DESCRIPTION
**Pull request by tremble**
##### SUMMARY

Update policy on existing keys when policy is passed as an option.

Fixes #59987

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

lib/ansible/modules/cloud/amazon/aws_kms.py

##### ADDITIONAL INFORMATION

@willthames PR as promised on Friday

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/aws_kms.py [hunk 1]
```diff
@@ -637,6 +637,28 @@ def update_key(connection, module, key):
...
+        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+            # If we can't fetch the current policy assume we're making a change
+            # Could occur if we have PutKeyPolicy without GetKeyPolicy
+            original_policy = {}
+        original_policy = json.loads(keyret['Policy'])
+        try:
+            new_policy = json.loads(policy)
+        except ValueError as e:
+            module.fail_json_aws(e, msg="Unable to parse new policy as JSON")
+        if json.dumps(original_policy, sort_keys=True) != json.dumps(new_policy, sort_keys=True) :
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/aws_kms.py]: extra space here before the colon

**Comment by willthames** [comment]
@tremble this change looks reasonable to me but needs the test suite updated to match (so that it would have caught this problem)

File: lib/ansible/modules/cloud/amazon/aws_kms.py (hunk 1 above)
Comment (tremble) [reply lib/ansible/modules/cloud/amazon/aws_kms.py]: removed

**Comment by tremble** [comment]
@willthames the integration tests appear to be broken.  This may be due to the older version of Boto3 I'm running (rhel 7), however, I'm seeing things like 

```
The full traceback is:
Traceback (most recent call last):
  File "/home/mchappel/.ansible/tmp/ansible-tmp-1565005313.24-218693380161805/AnsiballZ_aws_kms.py", line 139, in <module>
    _ansiballz_main()
  File "/home/mchappel/.ansible/tmp/ansible-tmp-1565005313.24-218693380161805/AnsiballZ_aws_kms.py", line 131, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/mchappel/.ansible/tmp/ansible-tmp-1565005313.24-218693380161805/AnsiballZ_aws_kms.py", line 68, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_aws_kms_payload_JEyiic/__main__.py", line 937, in <module>
  File "/tmp/ansible_aws_kms_payload_JEyiic/__main__.py", line 903, in main
KeyError: 'key_arn'
```

Which looks like some refactoring around aliases hasn't done what's expected.

**Comment by tremble** [comment]
See also #60206 for initial test fixes

**Comment by omonnig** [comment]
This PR looks reasonable, but do we care if the policy matches? I wasted way too much time trying to debug my policy before checking the module source and finding the root cause.  Just write the policy and be done with it.

How do we handle the BypassPolicyLockoutSafetyCheck?  It should default to False, only set to  True by the user.  This behavior can be dangerous, but allowed IMHO.

**Comment by willthames** [comment]
We do care if the policy matches, otherwise how do we know whether to report `changed` correctly?

**Comment by tremble** [comment]
@omonnig Am I understanding correctly that this PR **falsely** flagged your new policy as not changing anything?  Or are you just worried about the additional complexity?

As @willthames mentioned it's there so that "changed" can be correctly reported.  Personally I try to ensure that my playbooks are idempotent and "changed" only shows up when something's really changed.

re BypassPolicyLockoutSafetyCheck that would likely want to be an additional option, and for the sake of keeping PRs simple I'd rather not pull it into this PR.

**Comment by tremble** [comment]
@willthames Should I rebase this change on top of #60206 and see if I can test the shape / policy pieces as a part of this PR built on top of #60206 ?

**Comment by omonnig** [comment]
Sorry, @willthames, I was not thinking straight.  The Changed functionality is very important and should be included in this PR. 

I will put the BypassPolicyLockoutSafetyCheck in another PR.

**Comment by omonnig** [comment]
I was able to test against a live AWS account today.  The proposed policy equality test does not work when the policy is unchanged.  JSON IAM Policy returned from AWS does not return in the same key order every time, so the Policy must be sorted before comparison.  Found the snippet below, which works nicely.

# compare objects
# https://stackoverflow.com/questions/25851183/how-to-compare-two-json-objects-with-the-same-elements-in-a-different-order-equa
def ordered(obj):
    if isinstance(obj, dict):
        return sorted((k, ordered(v)) for k, v in obj.items())
    if isinstance(obj, list):
        return sorted(ordered(x) for x in obj)
    else:
        return obj

So the comparison changes to this:
        # if json.dumps(original_policy, sort_keys=True) != json.dumps(new_policy, sort_keys=True):
        if ordered(original_policy) != ordered(new_policy):

and the Changed flag functions as expected.

**Comment by willthames** [comment]
@omonnig we have `compare_policies` in module_utils/ec2.py already - so we probably just need to use it!

**Comment by willthames** [comment]
However, having said that and having a quick review of the code, it's complicated because AWS manipulates the policy, converting AWS principals to user ARNs and from role ARNs to unique role IDs (so that if a role is deleted and then a new role added with the same name, the new role won't be given those policies) and a lot of the `do_grant` code is simply reflecting that. It would be great to have a solution for that problem

**Comment by tremble** [comment]
@willthames @omonnig https://github.com/tremble/ansible/tree/aws_kms_keyid_only now uses compare_policies instead of json.dumps I'll update this PR once @jillr or @willthames can give me an idea which order we're going to try and merge everything in.

**Comment by tremble** [comment]
And just for the curious, the reason a simple json dump doesn't work is that it would appear Amazon mangle the Principal lists. My basic testing seems to show that when you add a new principal to the list then where-ever in the list you add it Amazon will add it to the end of the list.

**Review by jillr — APPROVED** [review]
Test suite passes for me, change looks good.  Thanks @tremble.

**Comment by willthames** [comment]
Looks like security-policy.json is missing a permission:

```
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_aws_kms_payload_gnmjc6sa/__main__.py", line 685, in update_key
    connection.put_key_policy(KeyId=key['key_id'], PolicyName='default', Policy=policy)
  File "/usr/local/lib/python3.6/dist-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python3.6/dist-packages/botocore/client.py", line 661, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (AccessDeniedException) when calling the PutKeyPolicy operation: User: arn:aws:iam::864163117881:user/AnsibleTestUser is not authorized to perform: kms:PutKeyPolicy on resource: arn:aws:kms:us-west-2:864163117881:key/57be8964-e790-44fb-afe1-2307e8febccd
```

**Comment by tremble** [comment]
@willthames added

**Comment by willthames** [comment]
Tests pass for me locally now, thanks @tremble!

**Pull request merged**